In [268]:
import pandas as pd
import numpy as np
import pickle
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
from sklearn.model_selection import KFold, cross_val_score



### Carregando variáveis previsoras e alvo

In [269]:
arq_alvo = open('../data/alvo.pkl', 'rb')
arq_previsores = open('../data/previsores.pkl', 'rb')
arq_previsores_esc = open('../data/previsores_esc.pkl', 'rb')
arq_previsores2 = open('../data/previsores2.pkl', 'rb')
arq_previsores2_esc= open('../data/previsores2_esc.pkl', 'rb')


In [270]:
alvo = pickle.load(arq_alvo)
previsores = pickle.load(arq_previsores)
previsores_esc = pickle.load(arq_previsores_esc)
previsores2 = pickle.load(arq_previsores2)
previsores2_esc= pickle.load(arq_previsores2_esc)

In [271]:
arq_alvo.close()
arq_previsores.close()
arq_previsores_esc.close()
arq_previsores2.close()
arq_previsores2_esc.close()

### Base de treino e teste

In [368]:
x_treino, x_teste, y_treino, y_teste = train_test_split(previsores2_esc, alvo, test_size=0.2, random_state=42)

In [273]:
x_treino.shape

(277, 11)

In [274]:
x_teste.shape

(70, 11)

In [275]:
y_treino.shape

(277,)

In [276]:
y_teste.shape

(70,)

# ALGORITMO NAVIE BAYES

In [277]:
from sklearn.naive_bayes import GaussianNB

In [278]:
naive = GaussianNB()
naive.fit(x_treino, y_treino);

### Avaliação do algoritmo

In [279]:
previsoes_naive = naive.predict(x_teste)
previsoes_naive

array([0, 1, 0, 0, 0, 0, 1, 1, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 1, 1, 0, 1,
       0, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 0, 1, 0, 0,
       0, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 0, 1, 0, 0, 1, 0, 0, 0, 0,
       1, 1, 1, 0])

In [280]:
y_teste

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 0, 0, 0,
       0, 0, 0, 1, 1, 0, 0, 1, 0, 0, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1,
       0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 0, 0, 0, 1, 0, 0, 0,
       1, 1, 0, 1])

In [281]:
accuracy_score(y_teste,previsoes_naive)

0.5428571428571428

In [282]:
print("Acurácia: %.2f%%" % (accuracy_score(y_teste,previsoes_naive) * 100.0))

Acurácia: 54.29%


In [283]:
confusion_matrix(y_teste, previsoes_naive)

array([[27, 25],
       [ 7, 11]])

In [284]:
print(classification_report(y_teste, previsoes_naive))

              precision    recall  f1-score   support

           0       0.79      0.52      0.63        52
           1       0.31      0.61      0.41        18

    accuracy                           0.54        70
   macro avg       0.55      0.57      0.52        70
weighted avg       0.67      0.54      0.57        70



### Análise dados de treino

In [285]:
previsoes_treino = naive.predict(x_treino)
previsoes_treino

array([1, 1, 1, 0, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 1, 0,
       1, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1,
       0, 1, 0, 0, 0, 1, 1, 0, 1, 1, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 1, 0,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 0, 1,
       0, 1, 0, 1, 0, 1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 0, 1, 0,
       1, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 0, 1, 0, 1,
       1, 1, 0, 0, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 1, 1,
       1, 1, 1, 1, 0, 1, 1, 1, 1, 0, 1, 0, 1, 1, 0, 0, 0, 1, 1, 1, 0, 1,
       1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 1, 1, 1, 0, 0, 1, 1,
       1, 0, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 0, 1, 0, 0, 1, 1, 1, 0, 0,
       1, 1, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 0, 0, 0,
       1, 1, 0, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 0, 1, 0, 1, 1,
       1, 1, 1, 0, 1, 0, 1, 1, 1, 1, 1, 0, 1])

In [286]:
print("Acurácia: %.2f%%" % (accuracy_score(y_treino,previsoes_treino) * 100.0))

Acurácia: 47.65%


### Validação Cruzada

In [287]:
# Separando os dados em folds 
kfold = KFold(n_splits=30, shuffle=True, random_state=5)

In [288]:
# Criando modelo
modelo = GaussianNB()
resultado = cross_val_score(modelo, previsores, alvo, cv=kfold)

print("Acurácia: %.2f%%" % (resultado.mean() * 100.0))

Acurácia: 53.89%


# ALGORITMO SVM

In [289]:
from sklearn.svm import SVC

In [290]:
svm = SVC(kernel='rbf', random_state=1, C=100)
svm.fit(x_treino, y_treino)

SVC(C=100, random_state=1)

In [291]:
previsoes_svm = svm.predict(x_teste)
previsoes_svm

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0])

In [292]:
print("Acurácia: %.2f%%" % (accuracy_score(y_teste,previsoes_svm) * 100.0))

Acurácia: 74.29%


In [293]:
confusion_matrix(y_teste, previsoes_svm)

array([[51,  1],
       [17,  1]])

In [294]:
print(classification_report(y_teste, previsoes_svm))

              precision    recall  f1-score   support

           0       0.75      0.98      0.85        52
           1       0.50      0.06      0.10        18

    accuracy                           0.74        70
   macro avg       0.62      0.52      0.47        70
weighted avg       0.69      0.74      0.66        70



### Análise dados de treino

In [295]:
previsoes_treino = svm.predict(x_treino)
previsoes_treino

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

In [296]:
print("Acurácia: %.2f%%" % (accuracy_score(y_treino,previsoes_treino) * 100.0))

Acurácia: 77.62%


In [297]:
confusion_matrix(y_treino, previsoes_treino)

array([[208,   1],
       [ 61,   7]])

### validação cruzada

In [298]:
# Separando os dados em folds 
kfold = KFold(n_splits=30, shuffle=True, random_state=5)

In [299]:
# Criando modelo
modelo = SVC(kernel='rbf', random_state=1, C=100)
resultado = cross_val_score(modelo, previsores2_esc, alvo, cv=kfold)

print("Acurácia: %.2f%%" % (resultado.mean() * 100.0))

Acurácia: 72.60%


# REGRESSÃO LOGÍSTICA

In [300]:
from sklearn.linear_model import LogisticRegression

In [301]:
logistica = LogisticRegression(random_state=500, max_iter=1, penalty='l2',
                               tol=0.0001, C=1, solver='lbfgs')

In [302]:
logistica.fit(x_treino,y_treino)

/home/arturdias/.local/lib/python3.10/site-packages/sklearn/linear_model/_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression(C=1, max_iter=1, random_state=500)

In [303]:
logistica.intercept_

array([-0.86532981])

In [304]:
logistica.coef_

array([[-0.10093479,  0.22295015, -0.11338169, -0.05444414, -0.01014379,
        -0.19551917,  0.17071048,  0.12378229, -0.12378229,  0.21804126,
        -0.1727175 ]])

In [305]:
previsoes_logistica = logistica.predict(x_teste)

In [306]:
print("Acurácia: %.2f%%" % (accuracy_score(y_teste,previsoes_logistica) * 100.0))

Acurácia: 72.86%


In [307]:
confusion_matrix(y_teste, previsoes_logistica)

array([[48,  4],
       [15,  3]])

In [308]:
accuracy_score(y_treino, previsoes_treino)

0.776173285198556

In [309]:
print(classification_report(y_teste, previsoes_logistica))

              precision    recall  f1-score   support

           0       0.76      0.92      0.83        52
           1       0.43      0.17      0.24        18

    accuracy                           0.73        70
   macro avg       0.60      0.54      0.54        70
weighted avg       0.68      0.73      0.68        70



### Validação Cruzada

In [310]:
# Separando os dados em folds 
kfold = KFold(n_splits=30, shuffle=True, random_state=5)

In [311]:
# Criando modelo
modelo = LogisticRegression(random_state=500, max_iter=100, penalty='l2',
                               tol=0.0001, C=1, solver='lbfgs')
resultado = cross_val_score(modelo, previsores2_esc, alvo, cv=kfold)

print("Acurácia: %.2f%%" % (resultado.mean() * 100.0))

Acurácia: 75.18%


# ALGORITMO KNN

In [312]:
from sklearn.neighbors import KNeighborsClassifier

In [313]:
knn = KNeighborsClassifier(n_neighbors=5, metric='minkowski', p=2)
knn.fit(x_treino, y_treino)

KNeighborsClassifier()

In [314]:
previsoes_knn = knn.predict(x_teste)

In [315]:
print("Acurácia: %.2f%%" % (accuracy_score(y_teste,previsoes_knn) * 100.0))

Acurácia: 78.57%


In [316]:
confusion_matrix(y_teste, previsoes_knn)

array([[50,  2],
       [13,  5]])

In [317]:
print(classification_report(y_teste, previsoes_knn))

              precision    recall  f1-score   support

           0       0.79      0.96      0.87        52
           1       0.71      0.28      0.40        18

    accuracy                           0.79        70
   macro avg       0.75      0.62      0.63        70
weighted avg       0.77      0.79      0.75        70



### Análise dados de treino

In [318]:
precisoes_treino= knn.predict(x_treino)

In [319]:
print("Acurácia: %.2f%%" % (accuracy_score(y_treino,previsoes_treino) * 100.0))

Acurácia: 77.62%


In [320]:
confusion_matrix(y_treino, previsoes_treino)

array([[208,   1],
       [ 61,   7]])

### Validação cruzada

In [321]:
# Separando os dados em folds 
kfold = KFold(n_splits=30, shuffle=True, random_state=5)

In [322]:
# Criando modelo
modelo = KNeighborsClassifier(n_neighbors=5, metric='minkowski', p=2)
resultado = cross_val_score(modelo, previsores2, alvo, cv=kfold)

print("Acurácia: %.2f%%" % (resultado.mean() * 100.0))

Acurácia: 72.93%


# Árvore de decsisão

In [323]:
from sklearn.tree import DecisionTreeClassifier

In [324]:
arvore = DecisionTreeClassifier(criterion='entropy', random_state=0, max_depth=2)
arvore.fit(x_treino, y_treino)

DecisionTreeClassifier(criterion='entropy', max_depth=2, random_state=0)

In [325]:
previsoes_arvore = arvore.predict(x_teste)

In [326]:
print("Acurácia: %.2f%%" % (accuracy_score(y_teste,previsoes_arvore) * 100.0))

Acurácia: 72.86%


In [327]:
confusion_matrix(y_teste, previsoes_arvore)

array([[49,  3],
       [16,  2]])

In [328]:
print(classification_report(y_teste, previsoes_arvore))

              precision    recall  f1-score   support

           0       0.75      0.94      0.84        52
           1       0.40      0.11      0.17        18

    accuracy                           0.73        70
   macro avg       0.58      0.53      0.51        70
weighted avg       0.66      0.73      0.67        70



### Análise dados de treino

In [329]:
previsoes_treino = arvore.predict(x_treino)

In [330]:
print("Acurácia: %.2f%%" % (accuracy_score(y_treino,previsoes_treino) * 100.0))

Acurácia: 77.62%


In [331]:
confusion_matrix(y_treino, previsoes_treino)

array([[200,   9],
       [ 53,  15]])

### Validação Cruzada

In [332]:
# Separando os dados em folds 
kfold = KFold(n_splits=30, shuffle=True, random_state=5)

In [333]:
# Criando modelo
modelo = DecisionTreeClassifier(criterion='gini', random_state=0)
resultado = cross_val_score(modelo, previsores2, alvo, cv=kfold)

print("Acurácia: %.2f%%" % (resultado.mean() * 100.0))

Acurácia: 67.17%


# ALGORITMO RANDOM FOREST

In [334]:
from sklearn.ensemble import RandomForestClassifier

In [335]:
random = RandomForestClassifier(n_estimators=100, criterion='entropy', random_state=0, max_depth=5)
random.fit(x_treino,y_treino)

RandomForestClassifier(criterion='entropy', max_depth=5, random_state=0)

In [336]:
previsoes_random = random.predict(x_teste)

In [337]:
print("Acurácia: %.2f%%" % (accuracy_score(y_teste, previsoes_random) * 100.0))

Acurácia: 75.71%


In [338]:
confusion_matrix(y_teste, previsoes_random)

array([[51,  1],
       [16,  2]])

In [339]:
print(classification_report(y_teste, previsoes_random))

              precision    recall  f1-score   support

           0       0.76      0.98      0.86        52
           1       0.67      0.11      0.19        18

    accuracy                           0.76        70
   macro avg       0.71      0.55      0.52        70
weighted avg       0.74      0.76      0.69        70



### Análise dados de treino

In [340]:
previsoes_treino = random.predict(x_treino)

In [341]:
print("Acurácia: %.2f%%" % (accuracy_score(y_treino, previsoes_treino) * 100.0))

Acurácia: 79.42%


In [342]:
confusion_matrix(y_treino, previsoes_treino)

array([[205,   4],
       [ 53,  15]])

### Validação cruzada

In [343]:
# Separando os dados em folds 
kfold = KFold(n_splits=30, shuffle=True, random_state=5)

In [344]:
# Criando modelo
modelo = RandomForestClassifier(n_estimators=100, criterion='entropy', random_state=0, max_depth=5)
resultado = cross_val_score(modelo, previsores2, alvo, cv=kfold)

print("Acurácia: %.2f%%" % (resultado.mean() * 100.0))

Acurácia: 74.39%


# ALGORITMO XGBoost

In [345]:
from xgboost import XGBClassifier

In [346]:
xg = XGBClassifier(max_depth=4, learning_rate= 0.05, n_estimators=180, 
                   objective='binary:logistic',random_state=3)

xg.fit(x_treino,y_treino)

XGBClassifier(base_score=None, booster=None, callbacks=None,
              colsample_bylevel=None, colsample_bynode=None,
              colsample_bytree=None, device=None, early_stopping_rounds=None,
              enable_categorical=False, eval_metric=None, feature_types=None,
              gamma=None, grow_policy=None, importance_type=None,
              interaction_constraints=None, learning_rate=0.05, max_bin=None,
              max_cat_threshold=None, max_cat_to_onehot=None,
              max_delta_step=None, max_depth=4, max_leaves=None,
              min_child_weight=None, missing=nan, monotone_constraints=None,
              multi_strategy=None, n_estimators=180, n_jobs=None,
              num_parallel_tree=None, random_state=3, ...)

In [347]:
previsoes_xg = xg.predict(x_teste)

In [348]:
print("Acurácia: %.2f%%" % (accuracy_score(y_teste, previsoes_xg) * 100.0))

Acurácia: 78.57%


In [349]:
confusion_matrix(y_teste, previsoes_xg)

array([[51,  1],
       [14,  4]])

### Análise dados de treino

In [350]:
previsoes_treino = xg.predict(x_treino)

In [351]:
print("Acurácia: %.2f%%" % (accuracy_score(y_treino, previsoes_treino) * 100.0))

Acurácia: 81.59%


In [352]:
confusion_matrix(y_treino, previsoes_treino)

array([[200,   9],
       [ 42,  26]])

### Validação cruzada

In [353]:
# Separando os dados em folds 
kfold = KFold(n_splits=30, shuffle=True, random_state=5)

In [354]:
# Criando modelo
modelo = XGBClassifier(max_depth=4, learning_rate= 0.05, n_estimators=180, 
                   objective='binary:logistic',random_state=3)
resultado = cross_val_score(modelo, previsores2, alvo, cv=kfold)

print("Acurácia: %.2f%%" % (resultado.mean() * 100.0))

Acurácia: 73.79%


# ALGORITMO LIGTHGBM

In [369]:
import lightgbm as lgb

In [370]:
dataset = lgb.Dataset(x_treino,y_treino)

In [371]:
param = {'num_leaves':150,
         'objective':'binary',
         'max_depth':3,
         'learning_rate':.05,
         'max_bin':200}

In [372]:
lgbm=lgb.train(param,dataset,num_boost_round=150)

[LightGBM] [Info] Number of positive: 68, number of negative: 209
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.021658 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 72
[LightGBM] [Info] Number of data points in the train set: 277, number of used features: 10
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.245487 -> initscore=-1.122827
[LightGBM] [Info] Start training from score -1.122827
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[

In [373]:
previsoes_lgbm = lgbm.predict(x_teste)

In [374]:
for i in range(0, previsoes_lgbm.shape[0]):
    if previsoes_lgbm[i]>= .5:
        previsoes_lgbm[i] = 1
    else:
        previsoes_lgbm[i] = 0

In [375]:
print("Acurácia: %.2f%%" % (accuracy_score(y_teste, previsoes_lgbm) * 100.0))

Acurácia: 75.71%


### Análise dados de treino

In [376]:
previsoes_treino = lgbm.predict(x_treino)

In [377]:
for i in range(0, previsoes_treino.shape[0]):
    if previsoes_treino[i]>= .5:
        previsoes_treino[i] = 1
    else:
        previsoes_treino[i] = 0

In [378]:
print("Acurácia: %.2f%%" % (accuracy_score(y_treino, previsoes_treino) * 100.0))

Acurácia: 78.70%


In [379]:
confusion_matrix(y_treino, precisoes_treino)

array([[190,  19],
       [ 39,  29]])

### Validação cruzada

In [380]:
# Separando os dados em folds 
kfold = KFold(n_splits=30, shuffle=True, random_state=5)

In [381]:
# Criando modelo
modelo = lgb.LGBMClassifier( num_leaves=150,objective='binary',max_depth=3,
                            learning_rate=.05, max_bin=200)
resultado = cross_val_score(modelo, previsores2_esc, alvo, cv=kfold)

print("Acurácia: %.2f%%" % (resultado.mean() * 100.0))

[LightGBM] [Info] Number of positive: 81, number of negative: 254
[LightGBM] [Info] Auto-choosing row-wise multi-threading, the overhead of testing was 0.000061 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 79
[LightGBM] [Info] Number of data points in the train set: 335, number of used features: 11
[LightGBM] [Info] [binary:BoostFromScore]: pavg=0.241791 -> initscore=-1.142885
[LightGBM] [Info] Start training from score -1.142885
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[

# ACURÁCIA DOS MODELOS
- Regressão Logística 75.18%
- Random Forest 74.39%
- XGBoost 73.79%
- LGBM 73.18%
- KNN 72.93%
- SVC 72.60%
- Árvore de decisão 67.17%
- Navie Bayes 53.89%